File: lib/writer.py
import json

from .driver.support.driver_support import DriverSupport


class Writer:
    """Writer object handles all file operations"""

    def __init__(self, f):
        """Writer constructor to initialize object

        :param File f: File object
        """
        self._f = f
        Writer._init_mimes()

    @staticmethod
    def _init_mimes():
        """Initialize MIME types data file"""
        with open("./data/mimes.json", "r") as f:
            mime_list = json.load(f)
            DriverSupport.MIMES = ','.join(mime_list)

    def write(self, files) -> None:
        """Write links to a file

        :param list files: list of file links
        :return: None
        """
        output = ''
        for file in files:
            output += f"{file}\n"
        self._f.write(output)

    def close(self) -> None:
        """Close File resource"""
        self._f.close()

    @staticmethod
    def read_input_file(file) -> list:
        """Read links from an file

        :param File file: file containing links ordered line by line
        :return: list of OD URLs
        """
        if file is not None:
            links = file.readline()
            urls = []
            while bool(links) or links != '\n':
                if links.endswith('\n'):
                    links = links[:len(links) - 1]
                    urls.append(links)
                    links = file.readline()
                else:
                    urls.append(links)
                    break

            file.close()
            return urls
        else:
            return []

File: app/src/main/java/xyz/zimtools/zyod/Writer.java
package xyz.zimtools.zyod;

import xyz.zimtools.zyod.args.Args;
import xyz.zimtools.zyod.assets.ODUrl;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Scanner;

/**
 * Handles file operations.
 */
public final class Writer {
    /**
     * Extract data from Mime file.
     * <p>
     * The extracted mime data will be transformed into a comma delimited format.
     * </p>
     *
     * @return all mime types in a comma separated format.
     */
    public static String readMimeFile() {
        try (InputStream input = AppConfig.getMimeStream()) {
            if (input == null) {
                throw new FileNotFoundException(AppConfig.MIME_FILE + " resource cannot " +
                        "be found!");
            }
            try (InputStreamReader reader = new InputStreamReader(input);
                 BufferedReader bufferedReader = new BufferedReader(reader)) {
                Optional<String> mimeTypes = bufferedReader.lines()
                        .map(line -> line.replaceAll(", +", ",")
                                .replace("[", "")
                                .replace("]", "")
                                .replace("\"", "")
                                .replace("\\", "\"")
                                .trim())
                        .reduce((acc, nextLine) -> acc + nextLine);
                return mimeTypes.orElse("");
            }
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
        return "";
    }

    /**
     * Read URLs from a file.
     * <p>
     * Retrieve links from a file. Each line in the file must contain one link to an OD to
     * operate on.
     * </p>
     * @param inputFile the input file
     * @return list of OD references
     */
    public static List<String> readInputFile(File inputFile) {
        List<String> urls = new ArrayList<>();
        try (Scanner scanner = new Scanner(inputFile)) {
            while (scanner.hasNext()) {
                String url = scanner.nextLine();
                urls.add(url);
            }
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
        return urls;
    }

    /**
     * Record {@link ODUrl}s to a file specified by {@link Args}.
     */
    public static void recordFiles(File outputFile, List<ODUrl> files) {
        try (FileWriter fileWriter = new FileWriter(outputFile, true);
             BufferedWriter bufferedWriter = new BufferedWriter(fileWriter)) {
            outputFile.createNewFile();
            for (ODUrl odUrl : files) {
                bufferedWriter.write(odUrl.toString());
                bufferedWriter.newLine();
            }
        } catch (IOException e) {
            Talker.warning(e.getMessage(), true);
        }
    }
}